pex/pip/tool.py

```diff
@@ -401,6 +405,36 @@ def _spawn_pip_isolated_job(
         )
         return Job(command=command, process=process, finalizer=finalizer)
 
+    @staticmethod
+    def _iter_build_configuration_options(build_configuration):
+        # type: (BuildConfiguration) -> Iterator[str]
+
+        # N.B.: BuildConfiguration maintains invariants that ensure --only-binary, --no-binary,
+        # --prefer-binary, --use-pep517 and --no-build-isolation are coherent.
+
+        if not build_configuration.allow_builds:
+            yield "--only-binary"
+            yield ":all:"
+        elif not build_configuration.allow_wheels:
+            yield "--no-binary"
+            yield ":all:"
+        else:
+            for project in build_configuration.only_wheels:
+                yield "--only-binary"
+                yield str(project)
+            for project in build_configuration.only_builds:
+                yield "--no-binary"
+                yield str(project)
+
+        if build_configuration.prefer_older_binary:
+            yield "--prefer-binary"
+
+        if build_configuration.use_pep517 is not None:
+            yield "--use-pep517" if build_configuration.use_pep517 else "--no-use-pep517"
```

Comment:
Q: what makes the pep517 option different from the other bool options?

_(realized I could probably reply to myself here by reading `pip install --help`)_
A: the difference aiui is that pep517 is used by default, so the option really is about disabling it.

Reply:
Thanks for digging a bit there, but I'm still crying a bit. Did you check out Pex's own docs?:
https://github.com/pantsbuild/pex/blob/015a077d3496205824491bf97136afd719277b12/pex/resolve/resolver_options.py#L206-L224

Assuming those are read, if they still don't make sense, it would be good to know so I can fix.

Reply:
I fixed that bad URL.

Reply:
Ah, my bad for not realizing the option was of course documented for pex as well (got tunnel visioned onto the `pip` arg..).

Yea, the pex option has a lot of information to digest. Reading it through a few times makes it almost clear. I wonder if there is a typo here:
```diff
---  If PEP-517 is forced (--use-pep517 is passed)
+++  If PEP-517 is forced (--force-pep517 is passed)
```

then again, I would expect that `--force-pep517` is the same as `--use-pep517` (based on my understanding of how `ArgumentParser` works), but the help has so many layers that I get the impression that they are not.. (also, that it reads `default: None` rather than `default: True` doesn't help which would align with the help text stating that it uses pep517 by default)
So, I feel there's a bit of language ... (failed to find the word I was looking for) here.

I think this option would be clearer using an enum value, but I guess the current state is due to upholding backwards compatibility..? In which case I think making the relationship between `--force-pep517` and `--use-pep517` clearer would help, and perhaps also mentioning that the default behavior is based on not providing a value for the option (e.g. explain that this flag is a tri-bool)

I can see why you cry a bit over this.

Reply:
Ok, so to sum up - I think you figured this out, but in case not:
+ The option is a None|True|False tri-state
+ The default is None
+ The option names are all just synonyms (with --no- prefixes negating)

So the default, None, means do the right thing: if there is a project.toml build system, use it; if not concoct one based on legacy setuptools build backend.

True means use (force) pep 517 always.

False means do not use pep 517.

I have no clue what use cases benefit from the True or False cases. I just exposed this Pip knob to not get between users and Pip. Perhaps my sin was in trying to make the Pip help different / more clear in up in Pex. I clearly failed there!

Reply:
Thanks. Yea, that summation aligns with my current understanding.